### PR TITLE
work on comparison

### PR DESCRIPTION
**Pull request by AviAvni**
fix https://github.com/RedisGraph/RedisGraph/issues/2435 https://github.com/RedisGraph/RedisGraph/issues/2497

### CODE REVIEW

File: src/filter_tree/filter_tree.c [hunk 1]
```diff
@@ -166,6 +166,11 @@ int _applyFilter(SIValue *aVal, SIValue *bVal, AST_Operator op) {
 		return (op == OP_NEQUAL);
 	}
 
+	if(disjointOrNull == COMPARED_NAN) {
+		/* The filter passes if we're testing for inequality, and fails otherwise. */
+		return (op == OP_NEQUAL);
```
Comment (OfirMos) [inline src/filter_tree/filter_tree.c:171]: @AviAvni  Missing test here, I guess it's pretty easy to test it.
Just a suggestion: I think you can merge this if with the one above.

File: src/value.c [hunk 2]
```diff
@@ -513,6 +513,10 @@ int SIValue_Compare(const SIValue a, const SIValue b, int *disjointOrNull) {
 		case T_BOOL:
 			return SAFE_COMPARISON_RESULT(a.longval - b.longval);
 		case T_DOUBLE:
```
Comment (OfirMos) [inline src/value.c:515]: I saw you've changed in Register_BooleanFuncs the params to SI_ALL, aren't we missing here some types like: T_DATE?

File: src/arithmetic/boolean_funcs/boolean_funcs.c [hunk 3]
```diff
@@ -147,6 +139,8 @@ SIValue AR_EQ(SIValue *argv, int argc) {
 	int disjointOrNull = 0;
 	int res = SIValue_Compare(a, b, &disjointOrNull);
 	if(disjointOrNull == COMPARED_NULL) return SI_NullVal();
+	if(disjointOrNull == COMPARED_NAN) return SI_BoolVal(false);
```
Comment (OfirMos) [inline src/arithmetic/boolean_funcs/boolean_funcs.c:142]: @AviAvni Just making sure, if both element are NAN we treat them as not equal, is this intended?
This question is also relevant to the rest of the functions.

**Review by OfirMos — CHANGES_REQUESTED** [review]
@AviAvni Nice, pls TAL at my comments.

File: src/filter_tree/filter_tree.c (hunk 1 above)
Comment (AviAvni) [reply src/filter_tree/filter_tree.c:177]: yes this is checked in the TCK tests

File: src/value.c (hunk 2 above)
Comment (AviAvni) [reply src/value.c:541]: T_DATE is not used now

File: src/arithmetic/boolean_funcs/boolean_funcs.c (hunk 3 above)
Comment (AviAvni) [reply src/arithmetic/boolean_funcs/boolean_funcs.c:144]: yes this is checked in the TCK tests

File: src/filter_tree/filter_tree.c (hunk 1 above)
Comment (AviAvni) [reply src/filter_tree/filter_tree.c:177]: which test missing?

Comment (AviAvni) [reply src/filter_tree/filter_tree.c:172]: test added thanks

File: src/value.h [hunk 4]
```diff
@@ -75,6 +75,7 @@ typedef enum {
 
 #define DISJOINT INT_MAX
 #define COMPARED_NULL INT_MIN
+#define COMPARED_NAN -1
```
Comment (swilly22) [inline src/value.h:78]: consider changing `COMPARED_NAN` from `-1` to `INT_MIN-1`

File: src/value.c [hunk 5]
```diff
@@ -539,6 +539,10 @@ int SIValue_Compare(const SIValue a, const SIValue b, int *disjointOrNull) {
...
+			if(isnan(a.doubleval) || isnan(b.doubleval)) {
+				if(disjointOrNull) *disjointOrNull = COMPARED_NAN;
+			}
+
 			return SAFE_COMPARISON_RESULT(a.doubleval - b.doubleval);
```
Comment (swilly22) [inline src/value.c:546]: Do we want to perform `SAFE_COMPARISON_RESULT` if either `a` or `b` are `NaN` ?

File: src/arithmetic/boolean_funcs/boolean_funcs.c [hunk 6]
```diff
@@ -79,11 +79,9 @@ SIValue AR_GT(SIValue *argv, int argc, void *private_data) {
 	if(disjointOrNull == COMPARED_NULL) {
 		// Comparisons with NULL values always return NULL.
 		return SI_NullVal();
-	} else if(disjointOrNull == DISJOINT) {
-		// Emit error when attempting to compare invalid types
-		Error_SITypeMismatch(b, SI_TYPE(a));
-		return SI_NullVal(); // The return doesn't matter, as the caller will check for errors.
 	}
+	if(disjointOrNull == COMPARED_NAN) return SI_BoolVal(false);
+	if(disjointOrNull == DISJOINT) return SI_NullVal();
```
Comment (swilly22) [inline src/arithmetic/boolean_funcs/boolean_funcs.c:84]: this line can be merged with: `if(disjointOrNull == COMPARED_NULL) {`
please add a comment explaining the situation similar to line 80

File: src/arithmetic/boolean_funcs/boolean_funcs.c [hunk 7]
```diff
@@ -79,11 +79,9 @@ SIValue AR_GT(SIValue *argv, int argc, void *private_data) {
 	if(disjointOrNull == COMPARED_NULL) {
 		// Comparisons with NULL values always return NULL.
 		return SI_NullVal();
-	} else if(disjointOrNull == DISJOINT) {
-		// Emit error when attempting to compare invalid types
-		Error_SITypeMismatch(b, SI_TYPE(a));
-		return SI_NullVal(); // The return doesn't matter, as the caller will check for errors.
 	}
+	if(disjointOrNull == COMPARED_NAN) return SI_BoolVal(false);
```
Comment (swilly22) [inline src/arithmetic/boolean_funcs/boolean_funcs.c:83]: please add a comment explaining the situation similar to line 80

File: src/arithmetic/boolean_funcs/boolean_funcs.c [hunk 8]
```diff
@@ -115,11 +111,9 @@ SIValue AR_LT(SIValue *argv, int argc, void *private_data) {
...
+	if(disjointOrNull == DISJOINT) return SI_NullVal();
```
Comment (swilly22) [inline src/arithmetic/boolean_funcs/boolean_funcs.c:116]: due to the reparative nature of this code, consider adding a macro:
```c
#define HANDLE_DISJOINT_CMP(disjointOrNull) \
 	if((disjointOrNull) == COMPARED_NULL || (disjointOrNull) == COMPARED_NAN) return SI_BoolVal(false); \
 	if((disjointOrNull) == DISJOINT) return SI_NullVal();
```

File: src/arithmetic/boolean_funcs/boolean_funcs.c [hunk 9]
```diff
@@ -247,30 +243,30 @@ void Register_BooleanFuncs() {
...
 	types = array_new(SIType, 2);
-	array_append(types, (SI_NUMERIC | T_STRING | T_BOOL | T_ARRAY | T_MAP | T_NULL));
-	array_append(types, (SI_NUMERIC | T_STRING | T_BOOL | T_ARRAY | T_MAP | T_NULL));
+	array_append(types, SI_ALL);
+	array_append(types, SI_ALL);
 	func_desc = AR_FuncDescNew("le", AR_LE, 2, 2, types, ret_type, true, true);
 	AR_RegFunc(func_desc);
 
-	types = array_new(SIType, 2);
+	types = array_new(SIType, 1);
```
Comment (swilly22) [inline src/arithmetic/boolean_funcs/boolean_funcs.c:269]: `types = array_new(SIType, 2);`

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/arithmetic/boolean_funcs/boolean_funcs.c (hunk 8 above)
Comment (AviAvni) [reply src/arithmetic/boolean_funcs/boolean_funcs.c:118]: it will be hard to debug

File: src/value.c (hunk 5 above)
Comment (AviAvni) [reply src/value.c:546]: it will be calculated to nan with is the expected result

**Pull request merged**